File: dbgen/p03_iot.py
import duckdb, numpy as np, sys, os, math
from datetime import datetime, timedelta
from concurrent.futures import ProcessPoolExecutor
from utils.synth_utils import batched_insert, run_parallel


def generate_sites_chunk(start, end, regions):
    size = end - start
    rng = np.random.default_rng(start)
    region_indices = rng.integers(0, len(regions), size)
    lats = rng.uniform(-60, 60, size)
    lons = rng.uniform(-180, 180, size)
    tz_list = ["UTC", "US/Eastern", "Europe/Berlin", "Asia/Tokyo"]
    tz_indices = rng.integers(0, len(tz_list), size)
    
    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            f"Site-{i}",
            regions[region_indices[idx]],
            round(float(lats[idx]), 4),
            round(float(lons[idx]), 4),
            tz_list[tz_indices[idx]],
        ))
    return rows

def generate_devices_chunk(start, end, NS, dtypes, base):
    size = end - start
    rng = np.random.default_rng(start)
    site_ids = rng.integers(1, NS + 1, size)
    dtype_indices = rng.integers(0, len(dtypes), size)
    model_letters = ['A', 'B', 'C']
    model_letter_indices = rng.integers(0, len(model_letters), size)
    model_numbers = rng.integers(1, 6, size)
    v_majors = rng.integers(1, 5, size)
    v_minors = rng.integers(0, 10, size)
    v_patches = rng.integers(0, 100, size)
    days_back = rng.integers(0, 731, size)
    active_probs = rng.random(size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            int(site_ids[idx]),
            dtypes[dtype_indices[idx]],
            f"Model-{model_letters[model_letter_indices[idx]]}{model_numbers[idx]}",
            f"v{v_majors[idx]}.{v_minors[idx]}.{v_patches[idx]}",
            (base - timedelta(days=int(days_back[idx]))).date(),
            bool(active_probs[idx] > 0.05),
        ))
    return rows

def generate_readings_chunk(start, end, ND, base):
    size = end - start
    rng = np.random.default_rng(start)
    device_ids = rng.integers(1, ND + 1, size)
    seconds_offset = rng.integers(0, 180 * 86400 + 1, size)
    temps = rng.normal(20, 8, size)
    humids = rng.uniform(20, 95, size)
    pressures = rng.normal(1013, 15, size)
    batteries = rng.integers(5, 101, size)
    rssis = rng.integers(-90, -29, size)
    error_probs = rng.random(size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        rows.append((
            i,
            int(device_ids[idx]),
            base + timedelta(seconds=int(seconds_offset[idx])),
            round(float(temps[idx]), 2),
            round(float(humids[idx]), 2),
            round(float(pressures[idx]), 2),
            int(batteries[idx]),
            int(rssis[idx]),
            bool(error_probs[idx] < 0.02),
        ))
    return rows

def generate_maintenance_logs_chunk(start, end, ND, base, actions):
    size = end - start
    rng = np.random.default_rng(start)
    device_ids = rng.integers(1, ND + 1, size)
    hours_offset = rng.integers(0, 4321, size)
    action_indices = rng.integers(0, len(actions), size)
    tech_ids = rng.integers(1, 21, size)
    note_action_indices = rng.integers(0, len(actions), size)

    rows = []
    for idx, i in enumerate(range(start, end)):
        action = actions[action_indices[idx]]
        rows.append((
            i,
            int(device_ids[idx]),
            base + timedelta(hours=int(hours_offset[idx])),
            action,
            f"Tech-{tech_ids[idx]}",
            f"Performed {actions[note_action_indices[idx]]} on device",
        ))
    return rows


def main():
    sf = float(sys.argv[1]) if len(sys.argv) > 1 else 1.0
    sf_adj = sf * 20.0
    NS = max(3, int(30 * sf_adj))
    ND = max(10, int(150 * sf_adj))
    NR = max(100, int(200000 * sf_adj))
    NML = max(5, int(500 * sf_adj))


    os.makedirs("data", exist_ok=True)
    con = duckdb.connect("data/warehouse.duckdb")

    con.execute("""
    DROP TABLE IF EXISTS maintenance_logs; DROP TABLE IF EXISTS readings;
    DROP TABLE IF EXISTS devices; DROP TABLE IF EXISTS sites;
    CREATE TABLE sites(site_id INTEGER PRIMARY KEY, name VARCHAR,
        region VARCHAR, latitude DOUBLE, longitude DOUBLE, timezone VARCHAR);
    CREATE TABLE devices(device_id INTEGER PRIMARY KEY, site_id INTEGER,
        device_type VARCHAR, model VARCHAR, firmware VARCHAR,
        installed_date DATE, is_active BOOLEAN);
    CREATE TABLE readings(reading_id BIGINT PRIMARY KEY, device_id INTEGER,
        ts TIMESTAMP, temperature_c DOUBLE, humidity_pct DOUBLE,
        pressure_hpa DOUBLE, battery_pct TINYINT, rssi_dbm SMALLINT,
        error_flag BOOLEAN);
    CREATE TABLE maintenance_logs(log_id INTEGER PRIMARY KEY, device_id INTEGER,
        log_ts TIMESTAMP, action VARCHAR, technician VARCHAR, notes VARCHAR);
    """)

    base = datetime(2023, 1, 1)
    regions = ["NA", "EU", "APAC", "LATAM"]
    dtypes = ["temperature", "humidity", "pressure", "multi", "air_quality"]
    actions = ["calibrate", "replace_battery", "firmware_update", "repair", "inspect"]

    cpu_count = os.cpu_count()
    with ProcessPoolExecutor(max_workers=cpu_count) as executor:
        batched_insert(con, "sites", ['site_id', 'name', 'region', 'latitude', 'longitude', 'timezone'],
                       run_parallel(executor, generate_sites_chunk, NS, regions))
        batched_insert(con, "devices", ['device_id', 'site_id', 'device_type', 'model', 'firmware', 'installed_date', 'is_active'],
                       run_parallel(executor, generate_devices_chunk, ND, NS, dtypes, base))
        batched_insert(con, "readings", ['reading_id', 'device_id', 'ts', 'temperature_c', 'humidity_pct', 'pressure_hpa', 'battery_pct', 'rssi_dbm', 'error_flag'],
                       run_parallel(executor, generate_readings_chunk, NR, ND, base))
        batched_insert(con, "maintenance_logs", ['log_id', 'device_id', 'log_ts', 'action', 'technician', 'notes'],
                       run_parallel(executor, generate_maintenance_logs_chunk, NML, ND, base, actions))


    con.close()
    print(f"p03 done: sites={NS} devices={ND} readings={NR}")

if __name__ == "__main__":
    main()

File: dbgen-rust/src/p03_iot.rs
use chrono::{Duration, NaiveDate};
use duckdb::{Connection, DuckdbConnectionManager};
use indicatif::{ProgressBar, ProgressStyle};
use r2d2::Pool;
use rand::prelude::*;
use rand::rngs::SmallRng;
use rand_distr::{Distribution, Normal};

pub fn run(sf: f64, pool: &mut Pool<DuckdbConnectionManager>) -> duckdb::Result<()> {
    let sf_adj = sf * 220.0;
    let ns = (30.0 * sf_adj).max(3.0) as usize;
    let nd = (150.0 * sf_adj).max(10.0) as usize;
    let nr = (200000.0 * sf_adj).max(100.0) as usize;
    let nml = (500.0 * sf_adj).max(5.0) as usize;

    let con = &pool.get().expect("couldn't get connection");

    con.execute_batch(
        "DROP TABLE IF EXISTS maintenance_logs; DROP TABLE IF EXISTS readings;
         DROP TABLE IF EXISTS devices; DROP TABLE IF EXISTS sites;
         CREATE TABLE sites(site_id INTEGER PRIMARY KEY, name VARCHAR,
             region VARCHAR, latitude DOUBLE, longitude DOUBLE, timezone VARCHAR);
         CREATE TABLE devices(device_id INTEGER PRIMARY KEY, site_id INTEGER,
             device_type VARCHAR, model VARCHAR, firmware VARCHAR,
             installed_date DATE, is_active BOOLEAN);
         CREATE TABLE readings(reading_id BIGINT PRIMARY KEY, device_id INTEGER,
             ts TIMESTAMP, temperature_c DOUBLE, humidity_pct DOUBLE,
             pressure_hpa DOUBLE, battery_pct TINYINT, rssi_dbm SMALLINT,
             error_flag BOOLEAN);
         CREATE TABLE maintenance_logs(log_id INTEGER PRIMARY KEY, device_id INTEGER,
             log_ts TIMESTAMP, action VARCHAR, technician VARCHAR, notes VARCHAR);",
    )?;

    let base_ts = NaiveDate::from_ymd_opt(2023, 1, 1)
        .unwrap()
        .and_hms_opt(0, 0, 0)
        .unwrap();
    let regions = ["NA", "EU", "APAC", "LATAM"];
    let dtypes = [
        "temperature",
        "humidity",
        "pressure",
        "multi",
        "air_quality",
    ];
    let actions = [
        "calibrate",
        "replace_battery",
        "firmware_update",
        "repair",
        "inspect",
    ];
    let tz_list = ["UTC", "US/Eastern", "Europe/Berlin", "Asia/Tokyo"];

    let pb = ProgressBar::new(4);
    pb.set_style(
        ProgressStyle::default_bar()
            .template("[{elapsed_precise}] {bar:40.cyan/blue} {pos}/{len} {msg}")
            .unwrap(),
    );

    // 1. Sites
    crate::generate_table_parallel(con, "sites", ns, &pb, "Generating sites...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let name = format!("Site-{}", i);
        let region = regions[rng.gen_range(0..regions.len())];
        let lat = ((rng.gen_range(-60.0..60.0) * 10000.0) as f64).round() / 10000.0;
        let lon = ((rng.gen_range(-180.0..180.0) * 10000.0) as f64).round() / 10000.0;
        let tz = tz_list[rng.gen_range(0..tz_list.len())];
        (i as i32, name, region, lat, lon, tz)
    })?;

    // 2. Devices
    crate::generate_table_parallel(con, "devices", nd, &pb, "Generating devices...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let site_id = rng.gen_range(1..=ns) as i32;
        let dtype = dtypes[rng.gen_range(0..dtypes.len())];
        let model = format!(
            "Model-{}{}",
            ['A', 'B', 'C'][rng.gen_range(0..3)],
            rng.gen_range(1..6)
        );
        let firmware = format!(
            "v{}.{}.{}",
            rng.gen_range(1..5),
            rng.gen_range(0..10),
            rng.gen_range(0..100)
        );
        let installed_date = (base_ts - Duration::days(rng.gen_range(0..731))).date();
        let is_active = rng.gen_bool(0.95);
        (
            i as i32,
            site_id,
            dtype,
            model,
            firmware,
            installed_date,
            is_active,
        )
    })?;

    // 3. Readings
    crate::generate_table_parallel(con, "readings", nr, &pb, "Generating readings...", |i| {
        let mut rng = SmallRng::seed_from_u64(i as u64);
        let device_id = rng.gen_range(1..=nd) as i32;
        let ts = base_ts + Duration::seconds(rng.gen_range(0..180 * 86400));
        let temp_dist = Normal::new(20.0, 8.0).unwrap();
        let temp = ((temp_dist.sample(&mut rng) * 100.0) as f64).round() / 100.0;
        let humid = ((rng.gen_range(20.0..95.0) * 100.0) as f64).round() / 100.0;
        let press_dist = Normal::new(1013.0, 15.0).unwrap();
        let press = ((press_dist.sample(&mut rng) * 100.0) as f64).round() / 100.0;
        let battery = rng.gen_range(5..=101) as i8;
        let rssi = rng.gen_range(-90..-29) as i16;
        let error = rng.gen_bool(0.02);
        (
            i as i64, device_id, ts, temp, humid, press, battery, rssi, error,
        )
    })?;

    // 4. Maintenance Logs
    crate::generate_table_parallel(
        con,
        "maintenance_logs",
        nml,
        &pb,
        "Generating maintenance logs...",
        |i| {
            let mut rng = SmallRng::seed_from_u64(i as u64);
            let device_id = rng.gen_range(1..=nd) as i32;
            let ts = base_ts + Duration::hours(rng.gen_range(0..4321));
            let action = actions[rng.gen_range(0..actions.len())];
            let tech = format!("Tech-{}", rng.gen_range(1..21));
            let note = format!(
                "Performed {} on device",
                actions[rng.gen_range(0..actions.len())]
            );
            (i as i32, device_id, ts, action, tech, note)
        },
    )?;

    pb.finish_with_message("p03_iot complete");

    Ok(())
}
